### PR TITLE
[New Connector] ServiceNow: Ping feature

### PR DESCRIPTION
**Pull request by parth-elastic**
#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:275]:
```suggestion
                "display": "textarea",
                "label": "Comma-separated list of services",
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:281]:
```suggestion
                "default_value": RETRIES,
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:283]:
```suggestion
                "label": "Retries per request",
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:193]: Why are we returning the data from `name` for valid services but `label` for invalid services?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:143]:
```suggestion
                            f"Cannot proceed due to unexpected response type '{response.headers["Content-Type"]}'; response type must begin with 'application/json'."
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:137]: Question: does this mean that ServiceNow is still processing the request, or that they have failed to?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:159]:
```suggestion
                    f"Retry attempt {retry} out of {self.retry_count} failed. Exception: {exception}."
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:167]: Question: Does this need to wait for the `.sleep` command above to close?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:29]: Question: I can't find these used anywhere. Will they be used in the future?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:70]: Question: Should we use the configuration value `concurrent_downloads` here? Or is this something else?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:79]: Question: will having `total=None` default to a specific timeout value?

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:66]: I think you can remove this explicit check and use `@cached_property`. Should result in the same functionality. You can probably remove the `self.session` field then also and only use `_get_session`.

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:43]: Why does the `ServiceNowClient` extend the `BaseDataSource`?

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:200]:
```suggestion
```

IMO we should remove doc strings, when they don't add any additional information, which isn't already present in the method name.

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:253]:
```suggestion
```

Same rationale as above, the doc string doesn't add any additional information.

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:302]: Should this be decorated with `@retryable`?

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:323]:
```suggestion
```

The framework already logs a similar information. We can remove that IMO.

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:328]:
```suggestion
```

Same as rationale as in the above comments.

File: connectors/sources/servicenow.py
Comment (timgrein) [inline connectors/sources/servicenow.py:336]:
```suggestion
```

This is also redundant IMO as `sync_job_runner` logs something similar before creating a sync.

File: connectors/sources/tests/test_servicenow.py
Comment (timgrein) [inline connectors/sources/tests/test_servicenow.py:18]: Could use an `AsyncMock()` instead of a custom class?

File: connectors/sources/tests/test_servicenow.py
Comment (timgrein) [inline connectors/sources/tests/test_servicenow.py:67]:
```suggestion
    with pytest.raises(ConfigurableFieldValueError):
```

IMO we should check for the specific exception when possible

**Review by timgrein — COMMENTED** [review]
Left some comments, I think we can remove some of the doc strings, if they don't provide any additional information, when compared to the name of the method.

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:137]: Yes, request was failed and returned empty response

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:29]: Yes, It will be used for downloading attachment in `get_docs`

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:70]: For not limiting client creation in the session

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:193]: It will be helpful in `get_docs` for preparing URLs with `name` & kept `label` for invalid services so that user can easily find out which entered services are invalid

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:297]: Retry mechanism is already handle internally in `_api_call`

File: connectors/sources/tests/test_servicenow.py
Comment (parth-elastic) [reply connectors/sources/tests/test_servicenow.py:18]: Custom class still be needed after extending `AsyncMock`. So created custom class with `aenter` & `aexit`

File: connectors/sources/servicenow.py
Comment (timgrein) [reply connectors/sources/servicenow.py:43]: I think it's better to just yield the docs from the client "raw" and serialize them in the `ServiceNowDataSource`. When we think of inheritance as a "is a" relationship it's a bit confusing IMO, when the client is also a data source.

**Comment by moxarth-elastic** [comment]
buildkite test this

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [reply connectors/sources/servicenow.py:68]: Right, but in the configuration there's a field called `concurrent_downloads`. Shouldn't it be used here instead of the const `MAX_CONCURRENCY_SUPPORT`?

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [reply connectors/sources/servicenow.py:134]: In that case, I suggest:

```suggestion
                            "Request to ServiceNow server returned an empty response."
```

File: connectors/sources/servicenow.py
Comment (navarone-feekery) [inline connectors/sources/servicenow.py:307]: > It will be helpful in get_docs for preparing URLs with name & kept label for invalid services so that user can easily find out which entered services are invalid

In a comment above we were discussing names vs labels, this error here is an example of what is confusing me. It tells the user two things:
- Service _labels_ x, y, and z are unavailable
- Service _names_ a, b, and c are available 

It's comparing different things (labels vs names), so I think it can lead to confusion. Can you rework this error message to be consistent?
For example, either:

`f"Services '{', '.join(invalid_service_labels)}' are not available. Available services are: '{', '.join(service_labels)}'"`

Or

`f"Services '{', '.join(invalid_service_names)}' are not available. Available services are: '{', '.join(service_names)}'"`

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:68]: If user enters `concurrent_downloads` value as 5 then it will restrict client creation in session only with 5 which may affect performance. Our source handles 10 clients at a time

Comment (navarone-feekery) [reply connectors/sources/servicenow.py:68]: Okay, so it's a different use-case for the `concurrent_downloads` configuration field. What is `concurrent_downloads` used for then, if not this? Is it being added later?

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py]: Got it 👍 Updated

File: connectors/sources/servicenow.py
Comment (timgrein) [reply connectors/sources/servicenow.py:44]: Could you elaborate a bit on the circular inheritance/concurrency problem with an example? Maybe we need to simply redesign the class structure to remove those issues. IMO we definitely shouldn't inherit from `BaseDataSource` in a client class.

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:68]: Yup, It will be used for handling concurrency for downloading attachment content

**Review by navarone-feekery — COMMENTED** [review]
Thanks for addressing my comments. LGTM, but I'll wait until @timgrein has finished his review to approve.

File: connectors/sources/servicenow.py
Comment (parth-elastic) [reply connectors/sources/servicenow.py:44]: Faced this issue while implementing `get_docs` but for now we can remove this inheritance and will redesign class structure according to your suggestion

Comment (timgrein) [reply connectors/sources/servicenow.py:44]: Great, thanks!

**Comment by moxarth-elastic** [comment]
buildkite test this

**Review by timgrein — APPROVED** [review]
LGTM, thank you for implementing the proposed changes

**Pull request merged**
